UnityProject/Assets/Scripts/Core/Lifecycle/CleanupUtil.cs

```diff
@@ -0,0 +1,351 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEditor;
+using System;
+using UI.Action;
+using Audio.Containers;
+
+public static class CleanupUtil
+{
+	/// <summary>
+	/// this methods scans the provided list and removes from it all elements, that are not marked as DontDestroyOnLoad. Will keep elements that are equal to null
+	/// <paramref name="targetExtractor"/> is just used to get the MonoBehaviour that is then being used to figure out whether or not element should be removed
+	/// </summary>
+	/// <param name="listInQuestion"> is a list that should be altered according to the rules provided by the method</param>
+	/// <param name="targetExtractor"> is a delegate that is transforming <typeparamref name="T"/> into a <c>UnityEngine.MonoBehaviour</c> that is later then removed from the <paramref name="listInQuestion"> if it's not marked as DontDestroyOnLoad </param>
+	public static int RidListOfSoonToBeDeadElements<T>(IList<T> listInQuestion, Func<T, UnityEngine.MonoBehaviour> targetExtractor)
+	{
+		if (listInQuestion == null)
+		{
+			return -1;
+		}
+
+		List<T> survivorList = new List<T>();
+
+		for (int i = 0, max = listInQuestion.Count; i < max; i++)
+		{
+			MonoBehaviour target = targetExtractor(listInQuestion[i]);
+
+			if (listInQuestion[i] != null && (target == null || target.gameObject.scene.buildIndex == -1))
+			{
+				survivorList.Add(listInQuestion[i]);
```

Comment:
We prefer to minimize nesting whenever possible, so check if the entry you're checking is null and do `continue`